### PR TITLE
CM-310: Create endpoint for changing the password

### PR DESCRIPTION
**Pull request by sniedzielski**
TICKET: https://openimis.atlassian.net/browse/CM-310

### CODE REVIEW

File: coremis_app_integration/endpoints/change_password.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+            response_data["result"] = {}
+            return Response(response_data, status=status.HTTP_200_OK)
+        else:
+            response = FAILED
+            response['result'] = _('The new password does not match the confirmation of new one')
+            return Response(response, status=status.HTTP_400_BAD_REQUEST)
+    else:
+        response = FAILED
+        response['result'] = _('Please provide correct old password')
+        return Response(response, status=status.HTTP_400_BAD_REQUEST)
```
Comment (dborowiecki) [inline coremis_app_integration/endpoints/change_password.py]: Some improvement ideas: 
```suggestion
    old_password = request.data.get('oldpassword', None)
    new_password = request.data.get('newpassword', None)
    confirm_password = request.data.get('confirmpassword', None)
    if new_password != confim_password:
            response = FAILED
            response['result'] = _('The new password does not match the confirmation of new one')
            return Response(response, status=status.HTTP_400_BAD_REQUEST)
            
    user = request.user
    authenticated_user = authenticate(username=user.username, password=old_password)
    
    if not authenticated_user:   
        response = FAILED
        response['result'] = _('Please provide correct old password')
        return Response(response, status=status.HTTP_400_BAD_REQUEST)

    user.set_password(new_password)
    user.save()
    response_data = SUCCESS
    response_data["result"] = {}
    return Response(response_data, status=status.HTTP_200_OK)
```

To simplify checks. I'd also suggest adding try catch block on database operation.

**Review by dborowiecki — CHANGES_REQUESTED** [review]

File: coremis_app_integration/endpoints/change_password.py (hunk 1 above)
Comment (sniedzielski) [reply coremis_app_integration/endpoints/change_password.py]: changed with adding try catch. Furthermore dataclass approach implemented here

**Comment by sniedzielski** [comment]
@dborowiecki Please re-check

File: coremis_app_integration/endpoints/change_password.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+from rest_framework import status
+
+from coremis_app_integration.authorization import CustomBasicAuthentication
+from coremis_app_integration.dataclasses import (
+    FailedResponse,
+    SuccessResponse
+)
+
+
+REGEX_PATTERN = r'^(?=.*[0-9])(?=.*[a-z])(?=.*[A-Z])(?=.*[@#$%^&!+=])(?=\S+$).{8,}$'
```
Comment (dborowiecki) [inline coremis_app_integration/endpoints/change_password.py:31]: Please add some comment about the requirement coming from this regex

```suggestion
"""
In summary, the regex enforces a string (likely a password) to:
    Be at least 8 characters long
    Contain at least one digit
    Contain at least one lowercase letter
    Contain at least one uppercase letter
    Contain at least one special character from the set @#$%^&!+=
    Not contain any whitespace characters
"""

REGEX_PATTERN = r'^(?=.*[0-9])(?=.*[a-z])(?=.*[A-Z])(?=.*[@#$%^&!+=])(?=\S+$).{8,}$'
```

**Review by dborowiecki — CHANGES_REQUESTED** [review]

**Pull request merged**